providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+        b = rotr64(b ^ c, 24);                                                 \
+        a = a + b + 2 * mul_lower(a, b);                                       \
+        d = rotr64(d ^ a, 16);                                                 \
+        c = c + d + 2 * mul_lower(c, d);                                       \
+        b = rotr64(b ^ c, 63);                                                 \
+    } while ((void)0, 0)
+
+# undef BLAKE2_ROUND
+# define BLAKE2_ROUND(v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11,        \
+                       v12, v13, v14, v15)                                     \
+    do {                                                                       \
+        G(v0, v4, v8, v12);                                                    \
+        G(v1, v5, v9, v13);                                                    \
+        G(v2, v6, v10, v14);                                                   \
+        G(v3, v7, v11, v15);                                                   \
+        G(v0, v5, v10, v15);                                                   \
+        G(v1, v6, v11, v12);                                                   \
+        G(v2, v7, v8, v13);                                                    \
+        G(v3, v4, v9, v14);                                                    \
+    } while ((void)0, 0)
+
+# undef BLAKE2_ROUND_COLUMN
+# define BLAKE2_ROUND_COLUMN(x, i)                                             \
+    do {                                                                       \
+        BLAKE2_ROUND(                                                          \
+            x[16 * i],      x[16 * i + 1],  x[16 * i + 2],  x[16 * i + 3],     \
+            x[16 * i + 4],  x[16 * i + 5],  x[16 * i + 6],  x[16 * i + 7],     \
+            x[16 * i + 8],  x[16 * i + 9],  x[16 * i + 10], x[16 * i + 11],    \
+            x[16 * i + 12], x[16 * i + 13], x[16 * i + 14], x[16 * i + 15]     \
+        );                                                                     \
+    } while ((void)0, 0)
+
+# undef BLAKE2_ROUND_ROW
+# define BLAKE2_ROUND_ROW(x, i)                                                \
+    do {                                                                       \
+        BLAKE2_ROUND(                                                          \
+            x[2 * i],      x[2 * i + 1],  x[2 * i + 16],  x[2 * i + 17],       \
+            x[2 * i + 32], x[2 * i + 33], x[2 * i + 48],  x[2 * i + 49],       \
+            x[2 * i + 64], x[2 * i + 65], x[2 * i + 80],  x[2 * i + 81],       \
+            x[2 * i + 96], x[2 * i + 97], x[2 * i + 112], x[2 * i + 113]       \
+        );                                                                     \
+    } while ((void)0, 0)
+
+typedef struct {
+    uint64_t v[ARGON2_QWORDS_IN_BLOCK];
+} block;
+
+typedef enum {
+    ARGON2_VERSION_10 = 0x10,
```

Comment:
Do you see any need for the version 10 variant?
If so add a setter - otherwise maybe just remove it..

Reply:
Setter added. Code-wise, the distinction is trivial enough ... we might as well support it.